chef/cookbooks/barclamp/libraries/barclamp_library.rb

```diff
@@ -417,21 +418,93 @@ def self.size_to_bytes(s)
+          @recipe_depedencies.fetch(recipe, [])
+        end
+      end
+    end
+
     class Config
       class << self
         attr_accessor :node
 
-        def load(group, barclamp, instance = nil)
-          # If no instance is specified, see if this node uses an instance of
-          # this barclamp and use it
+        def last_two_configs(group, barclamp, instance)
+          prev_cfg = load(group, barclamp, instance)
+          instance = infer_instance_from_cached_databag(group, barclamp, instance)
+          cur_cfg = Chef::Role.load "#{barclamp}-config-#{instance}"
+          return prev_cfg, cur_cfg.default_attributes[barclamp]
+        end
+
+        def loadattr(collection, attrlist)
+          begin
+            attrlist.each do |item|
+              unless collection.key?(item)
+                Chef::Log.info("[smart] #{item} is missing from collection")
+                return nil
+              end
+              collection = collection[item]
+            end
+          rescue NoMethodError
+            Chef::Log.info("[smart] collection did not respond to []/key? while looking for #{item}")
+            return nil
+          end
+          collection
+        end
+
+        def guess_instance(barclamp, instance)
           if instance.nil? && @node[barclamp] && @node[barclamp][:config]
             instance = @node[barclamp][:config][:environment]
           end
 
-          # Accept environments passed as instances
           if instance =~ /^#{barclamp}-config-(.*)/
             instance = $1
           end
+          instance
+        end
+
+        def infer_instance_from_cached_databag(group, barclamp, instance)
+          if instance.nil?
+            # try the "default" instance, and fallback on any existing instance
+            instance = "default"
+            unless @cache["groups"][group].fetch(instance, {}).key?(barclamp)
+              # sort to guarantee a consistent order
+              @cache["groups"][group].keys.sort.each do |key|
+                # ignore the id attribute from the data bag item, which is not
+                # an instance
+                next if key == "id"
+                if @cache["groups"][group][key].key?(barclamp)
+                  instance = key
+                  break
+                end
+              end
+            end
+          end
+          instance
+        end
+
+        def changes_to_apply?(depedency, group = "openstack", instance = nil)
+          barclamp = depedency.shift
+          instance = guess_instance(barclamp, instance)
+          prev_cfg, curr_cfg = last_two_configs(group, barclamp, instance)
+          old = loadattr(prev_cfg, depedency)
+          new = loadattr(curr_cfg, depedency)
+          # Chef::Log.info("[smart] loadattr prev #{depedency}, #{prev_cfg}")
+          # Chef::Log.info("[smart] loadattr curr #{depedency}, #{curr_cfg.inspect}")
+          # Chef::Log.info("[smart] comparision #{old}, #{new}")
+          old != new
```

Comment:
Does this really work on "non-leave" attributes? E.g. if you specify  ["glance"]["api"] as a dependency. ICBW but my guess is that this will always be `true` in that case. Even if the attribute values didn't change.

Reply:
your skepticism (thanks for that) got me thinking, hth 

```ruby
irb(main):019:0> a = {'glance'=>{'api'=>{'aport'=>1111, 'bport'=>2222}}}
=> {"glance"=>{"api"=>{"aport"=>1111, "bport"=>2222}}}

irb(main):020:0> b = {'glance'=>{'api'=>{'aport'=>1111}}}
=> {"glance"=>{"api"=>{"aport"=>1111}}}

irb(main):021:0> a == b
=> false

irb(main):022:0> a != b
=> true

irb(main):023:0> aa = a['glance']['api']
=> {"aport"=>1111, "bport"=>2222}

irb(main):024:0> bb = b['glance']['api']
=> {"aport"=>1111}

irb(main):025:0> aa != bb
=> true

irb(main):026:0> b = {'glance'=>{'api'=>{'aport'=>1111, 'bport'=>2222}}}
=> {"glance"=>{"api"=>{"aport"=>1111, "bport"=>2222}}}

irb(main):027:0> bb = b['glance']['api']
=> {"aport"=>1111, "bport"=>2222}

irb(main):028:0> aa != bb
=> false
```

Reply:
Hm, fascinating. And it even seems to work when comparing a `Hash` as returned by Chef::Role with a `Mash` as returned by Chef::DataBagItem.